### PR TITLE
Improve theme support when JS is disabled

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Little PR to follow the system's light/dark theme when JS is disabled.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Finally fixed CI. ^^'

**Comment by ehuss** [comment]
@notriddle Is this something you'd be willing to help review? No worries if not, I don't want to pressure anyone.

File: src/theme/css/noscript.css [hunk 1]
```diff
@@ -0,0 +1,7 @@
+.left-buttons button {
+    display: none;
+}
+
+.sidebar {
+    transition: transform 0.3s; /* Animation: slide away */
+}
```
Comment (notriddle) [inline src/theme/css/noscript.css:7]: The normal CSS file has this in it:

```css
body:not(.sidebar-resizing) .sidebar {
    transition: transform 0.3s; /* Animation: slide away */
}
```

Is there some reason why this doesn't work?

**Comment by GuillaumeGomez** [comment]
I think it'd be nice for mdBook to get more reviewers, so if needed, don't hesitate to ask me too for other PRs.

File: src/theme/index.hbs [hunk 2]
```diff
@@ -85,9 +88,7 @@
             var html = document.querySelector('html');
             html.classList.remove('{{ default_theme }}')
             html.classList.add(theme);
-            var body = document.querySelector('body');
-            body.classList.remove('no-js')
-            body.classList.add('js');
+            document.body.classList.add("js");
```
Comment (notriddle) [inline src/theme/index.hbs:91]: If you're using a `noscript.css` file, do you need this at all?

**Review by notriddle — CHANGES_REQUESTED** [review]
It looks like it's just code cleanup without any user-visible change, but your commit message sounds like it's saying that themes actually *work* better in noscript.

I don't think this PR allows you to actually switch themes when JS is disabled (I don't know how you'd even do that). It does allow you to see the menu bar, but the theme button isn't there.

The code itself seems okay, other than one question I asked here.

File: src/theme/css/noscript.css (hunk 1 above)
Comment (GuillaumeGomez) [reply src/theme/css/noscript.css:7]: I think this one can be removed indeed.

File: src/theme/index.hbs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/theme/index.hbs:91]: For theme handling in case JS is actually enabled: you don't want the system theme to be applied in this case by default.

**Comment by GuillaumeGomez** [comment]
> It looks like it's just code cleanup without any user-visible change, but your commit message sounds like it's saying that themes actually _work_ better in noscript.

It's now taking into account the system theme, which was not the case until now. So it is an improvement. ;)

> I don't think this PR allows you to actually switch themes when JS is disabled (I don't know how you'd even do that). It does allow you to see the menu bar, but the theme button isn't there.

It doesn't indeed.

**Comment by notriddle** [comment]
> It's now taking into account the system theme, which was not the case until now. So it is an improvement. ;)

Okay. Please say that in the PR description. Instead of just "Little PR to improve theme support when JS is disabled.", I would prefer something like "Little PR to follow the system's light/dark theme when JS is disabled."

File: src/theme/index.hbs (hunk 2 above)
Comment (notriddle) [reply src/theme/index.hbs:91]: Ah, I understand now. This is specifically used by `variables.css`.

**Comment by GuillaumeGomez** [comment]
Good idea, done!

**Comment by ehuss** [comment]
This seems like quite a large change compared to #2441. Why is this approach preferred over the one in #2441?

**Comment by GuillaumeGomez** [comment]
This PR removes some JS (which is why there are more changes, removing all `.no-js` items in CSS for example), introduces `noscript.css` which allows to have style specifically when JS is disabled.

Overall this PR is more complete I'd say.

File: src/theme/css/chrome.css [hunk 3]
```diff
@@ -107,7 +104,7 @@ a > .hljs {
     overflow: hidden;
     text-overflow: ellipsis;
 }
-.js .menu-title {
+.menu-title {
     cursor: pointer;
```
Comment (notriddle) [inline src/theme/css/chrome.css:108]: If I understand it correctly, this change will result in the :point_up_2: cursor being used on the menu title, even when javascript is disabled. I don't think clicking on the title actually does anything when javascript is disabled, so you'd need to do something to prevent that.

File: src/theme/css/noscript.css [hunk 4]
```diff
@@ -0,0 +1,3 @@
+.left-buttons button {
+    display: none;
+}
```
Comment (notriddle) [inline src/theme/css/noscript.css:3]: So, `noscript.css` is literally one CSS rule.

Is this change actually worth it? More specifically, do you expect to have a lot more JS-specific CSS than just this?

File: src/theme/css/chrome.css (hunk 3 above)
Comment (GuillaumeGomez) [reply src/theme/css/chrome.css:108]: Good point. We could emulate the same behaviour (scrolling to the top) without JS by replacing this element with `<a href="#">...</a>`. Should I do it in this PR?

File: src/theme/css/noscript.css (hunk 4 above)
Comment (GuillaumeGomez) [reply src/theme/css/noscript.css:3]: Might not be worth it currently. I didn't look at the rest yet but I think a lot more things could be done without JS. Anyway, gonna remove this file and move the rule into the main CSS file with `body:not(.js)` and should be fine.

**Comment by GuillaumeGomez** [comment]
Removed the `noscript.css` file. Diff is much smaller now.

File: src/theme/css/chrome.css (hunk 3 above)
Comment (notriddle) [reply src/theme/css/chrome.css:111]: I think that should be a separate PR.

File: src/theme/css/general.css [hunk 5]
```diff
@@ -14,6 +14,10 @@ html {
     -webkit-text-size-adjust: none;
 }
 
+* {
+    color: var(--fg);
+}
```
Comment (notriddle) [inline src/theme/css/general.css:19]: Okay, so, why is this needed?

File: src/theme/css/chrome.css (hunk 3 above)
Comment (GuillaumeGomez) [reply src/theme/css/chrome.css:111]: :+1:

File: src/theme/css/general.css (hunk 5 above)
Comment (GuillaumeGomez) [reply src/theme/css/general.css:19]: Without this change, it looks like this:

image

Comment (notriddle) [reply src/theme/css/general.css:19]: Yes, but why is it happening? This change seems like a workaround more than an actual fix.

Comment (GuillaumeGomez) [reply src/theme/css/general.css:19]: I can't figure out why. In the inspector, it says that this is the color in use, but for reasons I can't understand it's not being applied.

Comment (notriddle) [reply src/theme/css/general.css:19]: Here's a patch that fixes the problem in both Firefox and Ungoogled Chromium when I try it:

```diff
diff --git a/src/theme/css/chrome.css b/src/theme/css/chrome.css
index 9dbfe2c18..369d59ae0 100644
--- a/src/theme/css/chrome.css
+++ b/src/theme/css/chrome.css
@@ -91,7 +91,7 @@ body.sidebar-visible #menu-bar {
     display: flex;
     margin: 0 5px;
 }
-body:not(.js) .left-buttons button {
+html:not(.js) .left-buttons button {
     display: none;
 }
 
@@ -427,7 +427,7 @@ ul#searchresults span.teaser em {
     -ms-user-select: none;
     user-select: none;
 }
-body:not(.sidebar-resizing) .sidebar {
+html:not(.sidebar-resizing) .sidebar {
     transition: transform 0.3s; /* Animation: slide away */
 }
 .sidebar code {
diff --git a/src/theme/css/general.css b/src/theme/css/general.css
index 0e7443cc7..7670b087d 100644
--- a/src/theme/css/general.css
+++ b/src/theme/css/general.css
@@ -14,10 +14,6 @@ html {
     -webkit-text-size-adjust: none;
 }
 
-* {
-    color: var(--fg);
-}
-
 body {
     margin: 0;
     font-size: 1.6rem;
diff --git a/src/theme/css/variables.css b/src/theme/css/variables.css
index a30b97eba..12d1db7a3 100644
--- a/src/theme/css/variables.css
+++ b/src/theme/css/variables.css
@@ -112,7 +112,7 @@
     --copy-button-filter-hover: invert(36%) sepia(70%) saturate(503%) hue-rotate(167deg) brightness(98%) contrast(89%);
 }
 
-.light, body:not(.js) {
+.light, html:not(.js) {
     --bg: hsl(0, 0%, 100%);
     --fg: hsl(0, 0%, 0%);
 
@@ -258,7 +258,7 @@
 }
 
 @media (prefers-color-scheme: dark) {
-    body:not(.js) {
+    html:not(.js) {
         --bg: hsl(200, 7%, 8%);
         --fg: #98a3ad;
 
diff --git a/src/theme/index.hbs b/src/theme/index.hbs
index ea19458ec..99d304fc1 100644
--- a/src/theme/index.hbs
+++ b/src/theme/index.hbs
@@ -85,7 +85,7 @@
             var html = document.querySelector('html');
             html.classList.remove('{{ default_theme }}')
             html.classList.add(theme);
-            document.body.classList.add("js");
+            html.classList.add("js");
         </script>
 
         <input type="checkbox" id="sidebar-toggle-anchor" class="hidden">
```

Comment (GuillaumeGomez) [reply src/theme/css/general.css]: Wow. This is really impressive. Thanks a lot!

**Comment by GuillaumeGomez** [comment]
Applied your suggestions. :)

File: src/theme/css/chrome.css [hunk 6]
```diff
@@ -465,7 +464,7 @@ ul#searchresults span.teaser em {
     left: calc(var(--sidebar-resize-indicator-width) * -1);
     right: unset;
 }
-.js .sidebar .sidebar-resize-handle {
+.sidebar .sidebar-resize-handle {
```
Comment (notriddle) [inline src/theme/css/chrome.css:467]: This change results in the sidebar handle being visible and having a cursor, but it doesn't actually work.

File: src/theme/css/chrome.css [hunk 7]
```diff
@@ -427,8 +427,7 @@ ul#searchresults span.teaser em {
     -ms-user-select: none;
     user-select: none;
 }
-.no-js .sidebar,
-.js:not(.sidebar-resizing) .sidebar {
+html:not(.sidebar-resizing) .sidebar {
     transition: transform 0.3s; /* Animation: slide away */
```
Comment (notriddle) [inline src/theme/css/chrome.css:431]: The `sidebar-resizing` and `sidebar-visible` classes are currently set on the body element. I would probably move them to the html element, just like all the other utility classes, so that you don't have to memorize which is which.

Right now, you get a very strange rubber-band effect when resizing the sidebar.

**Review by notriddle — CHANGES_REQUESTED** [review]

File: src/theme/css/chrome.css (hunk 7 above)
Comment (GuillaumeGomez) [reply src/theme/css/chrome.css:431]: Good point, updated it.

File: src/theme/css/chrome.css (hunk 6 above)
Comment (GuillaumeGomez) [reply src/theme/css/chrome.css]: Ah indeed. Good catch!

**Comment by GuillaumeGomez** [comment]
Updated!

**Review by notriddle — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
@ehuss PR is now ready. :)

**Review by ehuss — APPROVED** [review]
Thanks!
